community/modules/scripts/htcondor-install/files/autoscaler.py

```diff
@@ -260,45 +259,59 @@ def scale(self):
         currentTarget = int(responseGroupInfo["targetSize"])
         print("Current MIG target size: " + str(currentTarget))
 
+        currentActions= responseGroupInfo["currentActions"]
+        n_being_born = currentActions["creating"] + \
+            currentActions["recreating"] + \
+            currentActions["starting"] + \
+            currentActions["restarting"] + \
+            currentActions["verifying"]
+
+        # count nodes that are either idle and have joined the pool or are
+        # still joining the pool from the MIG; there is some potential for
+        # overlap with the "verifying" state that will be addressed by
+        # only deleting nodes whose uptime exceeds this period of overlap
+        n_idle = len(idle_node_ads)
+        n_idle_total = n_being_born + n_idle
+        print(f"Idle nodes found: {n_being_born} birthing and {n_idle} idling")
+
+        # always keep size tending toward the minimum idle VMs requested
+        delta = min_hosts_for_queue - currentTarget
+        if delta < self.compute_instance_min_idle:
+            min_hosts_for_queue += self.compute_instance_min_idle - delta - n_idle_total
```

Comment:
I believe L280 could be substituted on to be re-written as:
```
step1: min_hosts_for_queue += self.compute_instance_min_idle - delta - n_idle_total
step2: min_hosts_for_queue = min_hosts_for_queue + self.compute_instance_min_idle - (min_hosts_for_queue - currentTarget) - n_idle_total
step3: min_hosts_for_queue = min_hosts_for_queue + self.compute_instance_min_idle - min_hosts_for_queue + currentTarget - n_idle_total
step4: min_hosts_for_queue = current_target + self.compute_instance_min_idle - n_idle_total
```

First, I would argue that step 4 as a statement transmits clearer meaning. Second, it seems odd to me that `min_hosts_for_queue` as calculated on L237 is not incorporated in this value.

Reply:
Quote from https://github.com/GoogleCloudPlatform/hpc-toolkit/pull/983#discussion_r1127092373 (resolved and tightly coupled to this conversation)

> This block (L278,279,280) feels is quite confusing.
>
> I don't understand why we have the if statement. What if delta=10 and compute_instance_min_idle=2. I don't understand why in this case we wouldn't want to bump up to account for min idle.
>
> It seems to me like we could just set new_target = min_hosts_for_queue + compute_instance_min_idle. This could be done without regard to what is currently idle.

Reply:
I did originally write the block as

```python
min_hosts_for_queue = current_target + self.compute_instance_min_idle - n_idle_total
```

and had trouble justifying each term. So both this and "why if?" are good questions. I've given some thought and seen it "in the wild" and have come to the following proposal. In pseudo-Python:

```python
# no if block
new_target = current_target + self.compute_instance_min_idle - n_idle_total + n_request_cpus_in_idle_jobs
```

Each term is individually justifiable, using some basic feedback loop concepts:

* `current_target`: a stable feedback loop ought to naturally stay at its current value in the absence of other signals
* `self.compute_instance_min_idle - n_idle_total`: a feedback signal restoring the target to whatever value is necessary to maintain `min_idle`
* `n_request_cpus_in_idle_jobs`: a feedback signal setting the target to the most compact set of CPUs that will meet the needs of idle jobs

@nick-stroud let me know if the summary above seems (a) better in terms of scaling accuracy (b) better in terms of the equation seeming "natural"

I have also reached a couple other conclusions:

* the counting of idle hosts and idle jobs should account for whether they have participated a negotiation cycle. i.e. an idle host or job that hasn't even attempted to match yet isn't a growth/contraction signal. I believe that this will help account for imperfect job packing onto machines
* right now, there is a bias toward deleting machines that already exist vs machines that are just being created and haven't yet joined the pool. the bias should probably go in the other direction.

Reply:
The new logic seems much clearer to me. Basically everything now focus on idle jobs and idle nodes.